pydoctor/epydoc/markup/epytext.py

```diff
@@ -1230,35 +1240,35 @@ def descr(self):
 ##                    SUPPORT FOR EPYDOC
 #################################################################
 
-def parse_docstring(docstring, errors):
+def parse_docstring(docstring: str, errors: List[ParseError]) -> ParsedDocstring:
     """
     Parse the given docstring, which is formatted using epytext; and
-    return a C{ParsedDocstring} representation of its contents.
+    return a L{ParsedDocstring} representation of its contents.
+
     @param docstring: The docstring to parse
-    @type docstring: C{string}
     @param errors: A list where any errors generated during parsing
         will be stored.
-    @type errors: C{list} of L{ParseError}
-    @rtype: L{ParsedDocstring}
     """
     tree = parse(docstring, errors)
     if tree is None:
         return ParsedEpytextDocstring(None, ())
 
+    tree_children = cast(List[Element], tree.children)
+
     fields = []
-    if tree.children and tree.children[-1].tag == 'fieldlist':
+    if tree_children and tree_children[-1].tag == 'fieldlist':
         # Take field list out of the document tree.
-        field_list = tree.children.pop()
+        field_list = tree_children.pop()
+        field_children = cast(List[Element], field_list.children)
 
-        for field in field_list.children:
+        for field in field_children:
             # Get the tag
-            tag = field.children[0].children[0].lower()
-            del field.children[0]
+            tag = cast(str, cast(Element, field.children.pop(0)).children[0]).lower()
```

Comment:
I guess that `tree_children` cast is needed because `parse` may return None .
But we we need the cast here and in other places. Shouldn't Element.children already have List[Elemenet] anotation?

Reply:
No, some elements have `str` children as well.

The design of this code isn't very friendly to type annotations: the `Element` class is very generic, but that means that the code using it contains a lot of assumptions about what kind of children different elements can contain. And those assumptions are coded in a way that mypy needs a lot of help to understand.

For now, I prefer to annotate the current design. But in the future, we could consider a few different directions for this code:
- instead of building an intermediate representation in a tree, route parse events to the code that generates Stan objects
- have separate `Element` subclasses for different elements, that put restrictions on the number and types of child nodes
- build an intermediate representation using `docutils` objects and let `docutils` convert that to HTML